_posts/2023-04-12-making-verification-compelling-visual-verification-feedback-for-dafny.markdown

```diff
@@ -0,0 +1,591 @@
+
+- "We did try to make a model out of our algorithms and prove properties, but, well
+  only that senior mathematician on another specialized team can understand and maintain it, and they are retiring soon..."
+- "My developers have tried using modeling tools themselves but it's like... manipulating mathematics is discouraging..."
+
+If only it doing mathematics at the software level was engaging and compelling... and then you remember Dafny and its new features you will have read about in this blog post.
+
+### What?
+
+I'll explain how mathematical reasoning of programs can be made _more_ engaging by increasing the amount of negative, contextual and especially positive feedback in Dafny,
+The success was to the point point where a user spontaneously said:
+<div style="text-align:center;height:2.2em;font-style:italic">
+"It's ridiculous how good all green gutter feels"
+</div>
+
+After asking them if we could quote that sentence, another user added _"If he won’t, I say it for you to use"_. A third one said:
+
+<div style="text-align:center;height:2.2em;font-style:italic">
+"Oh these little icons on the left, they are super useful"
+</div>
+Isn't that a super cute testimony?
+
+## Dafny
+
+### Quick story of Dafny
+
+Dafny started as an imperative programming language with built-in support for verification in 2009, but you probably haven't heard a lot about it in the news. It's because it's in a special niche: It is thought for programmers, but designed with verification as the first goal in mind.
+
+Dafny has been successfully used in several industrial and academics utilizations.
+With its Java-like syntax and good type inference, Dafny has [2k stars on GitHub](https://github.com/dafny-lang/dafny).
+Dafny is being used both academic circles to [teach formal reasoning](http://www.doc.ic.ac.uk/~scd/Dafny_Material/Lectures.pdf), as well as in industrial projects like [IronFleet](https://github.com/microsoft/Ironclad/blob/main/ironfleet/README.md) and [IronClad projects](https://github.com/microsoft/Ironclad/blob/main/ironfleet/README.md), as part of the [AWS Encryption SDK](https://github.com/aws/aws-encryption-sdk-dafny) and as the model to verify the [AWS permission engine](https://aws.amazon.com/verified-permissions/).
+
+Under the hood, Dafny reasons about programs, and can not only prove the absence of run-time exceptions, but also invariants and specific properties relating inputs and outputs.
+
+### A quick introduction to Dafny
+
+By now, you might want to know what does it mean to reason about programs _mathematically_ to avoid bugs that tests might not cover. It's your lucky day, I made an animation to explain the process. Here is an example of an unoptimized program that computes the euclidian division in Dafny. To go to the next step, click "Next".
+
+<div class="step-through">
+<div class="step">
+<div class="file" name="euclidian.dfy">
```

Comment:
> We could pull this out, too.

What do you mean pull it out? HTML does not have a mechanism to include other html files.

Reply:
Isn't that what an iframe does?

Reply:
I meant, to include html code raw.

I would not like to use iframes. Let me explain you why
- I store the current animation step on a location;hash so that if someone refreshes the page, they are still in the same step. Also, useful for sharing. iframes have their own URL and location.hash which is not shareable. Starting to write code to interface the two is everything but simple refactoring.
- We already saw that iframes did not work in the context of the index trying to display the blog posts.-